### ISSUE TITLE
Switch zone root filesystem is not mounted due to name conflicts

### ISSUE THREAD

**Issue by bnaecker** [oxidecomputer/omicron#4269]
This is similar to #4203. On at least `madrid` (`BRM44220001`) and the Dogfood rack sled 14 (`BRM42220051`), the switch zone dataset `rpool/zone` is not mounted, apparently due to a name conflict:

```bash
BRM44220001 # zfs list -r rpool/zone
NAME         USED  AVAIL     REFER  MOUNTPOINT
rpool/zone    24K  3.02G       24K  /zone
BRM44220001 # zfs list -o name,canmount,mounted,mountpoint rpool/zone
NAME        CANMOUNT  MOUNTED  MOUNTPOINT
rpool/zone        on       no  /zone
BRM44220001 # zfs get -Hpo value name /zone
rpool/ROOT/ramdisk
```

Based on some comments in the sled-agent code, [here](https://github.com/oxidecomputer/omicron/blob/c6955a5a0452c958059ae1de9376389c0286c14e/illumos-utils/src/zfs.rs#L12) and [here](https://github.com/oxidecomputer/omicron/blob/c6955a5a0452c958059ae1de9376389c0286c14e/sled-agent/src/services.rs#L2867), we indeed want this dataset to be part of the ramdisk. But we also clearly expect it to be mounted at `/zone`, which it is definitely not.

As with #4203, I'm not sure where the directory `/zone` is created in the sled-agent.

**Comment by bnaecker** [comment]
Ok, I _think_ I've traced this one down. Here are my notes during debugging, which are a bit rough but show the whole process.


The switch zone root FS is _supposed_ to be `rpool/zone` which ought to be mounted at `/zone`. But looks like the call to `zoneadm install` is creating a directory there before ZFS can mount it.

```
BRM44220001 # zfs get all rpool/zone
NAME        PROPERTY              VALUE                  SOURCE
rpool/zone  type                  filesystem             -
rpool/zone  creation              Fri Oct 13  0:22 2023  -
rpool/zone  used                  24K                    -
rpool/zone  available             3.25G                  -
rpool/zone  referenced            24K                    -
rpool/zone  compressratio         1.00x                  -
rpool/zone  mounted               no                     -
rpool/zone  quota                 none                   default
rpool/zone  reservation           none                   default
rpool/zone  recordsize            128K                   default
rpool/zone  mountpoint            /zone                  local
rpool/zone  sharenfs              off                    default
rpool/zone  checksum              on                     default
rpool/zone  compression           gzip-9                 inherited from rpool
rpool/zone  atime                 on                     default
rpool/zone  devices               on                     default
rpool/zone  exec                  on                     default
rpool/zone  setuid                on                     default
rpool/zone  readonly              off                    default
rpool/zone  zoned                 on                     local
rpool/zone  snapdir               hidden                 default
rpool/zone  aclmode               discard                default
rpool/zone  aclinherit            restricted             default
rpool/zone  createtxg             25274                  -
rpool/zone  canmount              on                     default
rpool/zone  xattr                 on                     default
rpool/zone  copies                1                      default
rpool/zone  version               5                      -
rpool/zone  utf8only              off                    -
rpool/zone  normalization         none                   -
rpool/zone  casesensitivity       sensitive              -
rpool/zone  vscan                 off                    default
rpool/zone  nbmand                off                    default
rpool/zone  sharesmb              off                    default
rpool/zone  refquota              none                   default
rpool/zone  refreservation        none                   default
rpool/zone  guid                  14266145686336881846   -
rpool/zone  primarycache          none                   inherited from rpool
rpool/zone  secondarycache        all                    default
rpool/zone  usedbysnapshots       0                      -
rpool/zone  usedbydataset         24K                    -
rpool/zone  usedbychildren        0                      -
rpool/zone  usedbyrefreservation  0                      -
rpool/zone  logbias               latency                default
rpool/zone  dedup                 off                    default
rpool/zone  mlslabel              none                   default
rpool/zone  sync                  standard               default
rpool/zone  dnodesize             legacy                 default
rpool/zone  refcompressratio      1.00x                  -
rpool/zone  written               24K                    -
rpool/zone  logicalused           12K                    -
rpool/zone  logicalreferenced     12K                    -
rpool/zone  filesystem_limit      none                   default
rpool/zone  snapshot_limit        none                   default
rpool/zone  filesystem_count      none                   default
rpool/zone  snapshot_count        none                   default
rpool/zone  redundant_metadata    all                    default
rpool/zone  special_small_blocks  0                      default
rpool/zone  encryption            off                    default
rpool/zone  keylocation           none                   default
rpool/zone  keyformat             none                   default
rpool/zone  pbkdf2iters           0                      default
```

Note that its not mounted, and that there is already something at `/zone`:

```
BRM44220001 # ls /zone
BRM44220001 #
```

I'm tryin to understand how this happens, by restarting the sled agent and tracing the creation of that dir:

```
BRM44220001 # dtrace -n 'syscall::mkdir:entry { this->path = copyinstr(arg0); if (this->path == "/zone") { printf("%s\n", this->path); ustack(); printf("%s\n", curpsinfo->pr_psargs); exit(0); }  }'
dtrace: description 'syscall::mkdir:entry ' matched 5 probes
CPU     ID                    FUNCTION:NAME
 26  18433                      mkdir:entry /zone

              libc.so.1`syscall+0x5
              libgen.so.1`mkdirp+0xa7
              zoneadm`validate_zonepath+0x137
              zoneadm`verify_details+0x1f7
              zoneadm`install_func+0x3ed
              zoneadm`parse_and_run+0x2f
              zoneadm`main+0x2d5
              zoneadm`_start_crt+0x9a
              zoneadm`_start+0x1a
/usr/sbin/zoneadm -z oxz_switch install /opt/oxide/switch.tar.gz /opt/oxide/ove


```

Ok, so that's a call in the sled agent to `zoneadm install` the switch zone. That happens [here](https://github.com/oxidecomputer/omicron/blob/2af5ec9fd63495a3d0aa90f7d4b1cd1b9eae17ff/illumos-utils/src/zone.rs#L356). The configuration for that zone is:

```
BRM44220001 # zonecfg -z oxz_switch info
zonename: oxz_switch
zonepath: /zone/oxz_switch
brand: omicron1
autoboot: false
bootargs:
pool:
limitpriv: default,sys_dl_config
scheduling-class:
ip-type: exclusive
hostid:
fs-allowed:
net:
    address not specified
    allowed-address not specified
    physical: oxBootstrap0
    defrouter not specified
net:
    address not specified
    allowed-address not specified
    physical: oxControlService0
    defrouter not specified
net:
    address not specified
    allowed-address not specified
    physical: tfpkt0
    defrouter not specified
device:
    match: /dev/tofino/1
```

So the call to `zoneadm` appears to be creating that directory because it's (1) part of the root filesystem, and (2) doesn't yet exist. So the question is why does that not exist? The ZFS dataset appears to exist, but presumably can't be mounted because that directory exists first?

Well, it looks like the call to `zfs create rpool/zone` happens before the call to `mkdir` initiated by the `zoneadm install` call:

```
BRM44220001 # dtrace -n 'syscall::mkdir:entry { this->path = copyinstr(arg0); if (this->path == "/zone") { printf("%s\n", this->path); ustack(); printf("%s\n", curpsinfo->pr_psargs); }  }' -n 'proc:::exec-success /(execname == "zfs") && (strstr(curpsinfo->pr_psargs, "create") != NULL)/ { print(*curpsinfo); }'
dtrace: description 'syscall::mkdir:entry ' matched 5 probes
dtrace: description 'proc:::exec-success ' matched 1 probe
CPU     ID                    FUNCTION:NAME
112  19095         exec_common:exec-success psinfo_t {
    int pr_flag = 0
    int pr_nlwp = 0x1
    pid_t pr_pid = 0x6939
    ...
    char [16] pr_fname = [ "zfs" ]
    char [80] pr_psargs = [ "/usr/sbin/zfs create -o zoned=on -o mountpoint=/zone rpool/zone" ]
    int pr_wstat = 0
    ...
}
103  18433                      mkdir:entry /zone

              libc.so.1`syscall+0x5
              libgen.so.1`mkdirp+0xa7
              zoneadm`validate_zonepath+0x137
              zoneadm`verify_details+0x1f7
              zoneadm`install_func+0x3ed
              zoneadm`parse_and_run+0x2f
              zoneadm`main+0x2d5
              zoneadm`_start_crt+0x9a
              zoneadm`_start+0x1a
/usr/sbin/zoneadm -z oxz_switch install /opt/oxide/switch.tar.gz /opt/oxide/ove


```

So, the `zfs create` operation is first, then the call to `mkdir(2)`. Why does the mkdir call happen at all? First, does it _succeed_? Maybe the `zoneadm` command is trying to create it, but it's already there?

```
BRM44220001 # dtrace -n 'syscall::mkdir:entry { this->path = copyinstr(arg0); if (this->path == "/zone") { printf("%s\n", this->path); ustack(); printf("%s\n", curpsinfo->pr_psargs); self->creating = 1;}  }' -n 'syscall::mkdir:return /self->creating/ { printf("retval = %d\n", arg1); self->creating = 0; }'  -n 'proc:::exec-success /(execname == "zfs") && (strstr(curpsinfo->pr_psargs, "create") != NULL)/ { print(*curpsinfo); }'
...
  1  18433                      mkdir:entry /zone

              libc.so.1`syscall+0x5
              libgen.so.1`mkdirp+0xa7
              zoneadm`validate_zonepath+0x137
              zoneadm`verify_details+0x1f7
              zoneadm`install_func+0x3ed
              zoneadm`parse_and_run+0x2f
              zoneadm`main+0x2d5
              zoneadm`_start_crt+0x9a
              zoneadm`_start+0x1a
/usr/sbin/zoneadm -z oxz_switch install /opt/oxide/switch.tar.gz /opt/oxide/ove

  1  18434                     mkdir:return retval = 0
```

Nope, `mkdir` returns 0, so the directory did not exist when that function was called. I don't believe there's any kind of asynchronicity between when a filesystem is created and when it's mounted. I tried on my own machine, and `zfs_mount()` is called before the `zfs create` command returns.

**BUT!** Note the actual command we're running to create the dataset:

```
/usr/sbin/zfs create -o zoned=on -o mountpoint=/zone rpool/zone
```

That creates a zoned filesystem, and from the man pages:

> A zoned file system cannot be mounted in the global
 zone, since the zone administrator might have to set the mount point to
 an unacceptable value.

So it appears we're mounting the switch zone filesystem as zoned, and then expecting it to have a GZ mountpoint. Zoned filesystems are really intended to be delegated into the zone, e.g. by using `zonecfg add fs`. That allows the filesystem to be managed through `zfs(8)` from _within_ the zone. That's not really what we want to do here. The root filesystem isn't something we want programs in the zone to manage. Further, we can see the man pages clearly state that this precludes mounting in the GZ, since the mountpoint may not even exist in the GZ.

And sure enough:

https://github.com/oxidecomputer/omicron/blob/2af5ec9fd63495a3d0aa90f7d4b1cd1b9eae17ff/sled-agent/src/bootstrap/pre_server.rs#L370-L387

This dataset should be created without the `zoned=on` property. I've not verified, but it seems plausible that this is the same cause underlying #4203. We should look at the calls to `ensure_filesystem` that create those, and make sure they're not zoned as well.

**Comment by bnaecker** [comment]
I've changed that call to `zoned = false`, and now:

```
BRM44220001 # zfs list -Hpo name,mountpoint,mounted,zoned rpool/zone
rpool/zone	/zone	yes	off
```

I'll further test with the actual code I wanted to verify now.

**Comment by bnaecker** [comment]
Closed by 11bd01faa37263bf00b8e2af7a897882a93fbbb7
